### PR TITLE
Improve remote image updates in clusters

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

File: lxd/db/images.go [hunk 1]
```diff
@@ -1060,6 +1060,27 @@ func (c *Cluster) GetPoolsWithImage(imageFingerprint string) ([]int64, error) {
 	return poolIDs, nil
 }
 
+// GetPoolsWithRemoteImage gets the IDs of all storage pools on which a given image exists.
```
Comment (stgraber) [inline lxd/db/images.go:1063]: Probably worth expending a bit on this comment.

Looks like you're getting a list of pool IDs for pools backed by remote storage which hold a particular image.

File: lxd/images.go [hunk 2]
```diff
@@ -1153,6 +1155,61 @@ func autoUpdateImagesInProject(ctx context.Context, d *Daemon, project string) e
 		return errors.Wrap(err, "Unable to retrieve the list of images")
 	}
 
+	var nodes []api.ClusterMember
+
+	// Wait up to 30 seconds for all nodes to get online.
```
Comment (stgraber) [inline lxd/images.go:1160]: Why is all this logic needed?

I would have expected us to spread the image update over all online nodes by simply fetching the current nodes list, removing the offline node and then picking a "random" index in there for each image. If a node comes back online, then things will get reshuffled, but that should be fine.

File: lxd/images.go [hunk 3]
```diff
@@ -1164,6 +1221,34 @@ func autoUpdateImagesInProject(ctx context.Context, d *Daemon, project string) e
 			continue
 		}
 
+		// In theory, pools should contain at most 1 value. That is because the call to GetImage
+		// will fail if there are multiple matches.
+		// We intentionally ignore the error.
+		pools, _ := d.cluster.GetPoolsWithRemoteImage(fingerprint)
+
+		if len(nodes) > 1 && len(availableNodeIDs) > 0 && len(pools) == 1 {
+			isRemoteStorage, _ := d.cluster.IsRemoteStorage(pools[0])
```
Comment (stgraber) [inline lxd/images.go:1230]: I'm confused, doesn't `GetPoolsWithRemoteImage` already do that? From what I remember of that function, you were specifically filtering on `node_id IS NULL` which would only ever be true of remote storage pools.

File: lxd/images.go [hunk 4]
```diff
@@ -1164,6 +1221,34 @@ func autoUpdateImagesInProject(ctx context.Context, d *Daemon, project string) e
...
+				hash := fnv.New64a()
+				_, err = io.WriteString(hash, seed)
+				if err != nil {
+					continue
+				}
+
+				r := rand.New(rand.NewSource(int64(hash.Sum64())))
+				selectedNodeID := availableNodeIDs[r.Int63n(int64(len(availableNodeIDs)))]
+
+				// Don't snapshot, if we're not the chosen one.
```
Comment (stgraber) [inline lxd/images.go:1245]: That comment sounds wrong :)

**Comment by stgraber** [comment]
Looks like this is still very much a work in progress, so reset it to draft state :)

File: lxd/images.go (hunk 2 above)
Comment (monstermunchkin) [reply lxd/images.go:1160]: The images are also refreshed on startup. I noticed that all cluster nodes are marked as offline when this code in run and take some time to appear to be online. That's the reason I added this logic.

File: lxd/images.go (hunk 3 above)
Comment (monstermunchkin) [reply lxd/images.go:1230]: You're right. This check is unnecessary.

File: lxd/db/images.go [hunk 5]
```diff
@@ -29,6 +29,7 @@ import (
 //go:generate mapper stmt -p db -e image objects-by-Project-and-Fingerprint-and-Public
 //go:generate mapper stmt -p db -e image objects-by-Fingerprint
 //go:generate mapper stmt -p db -e image objects-by-Cached
+//go:generate mapper stmt -p db -e image objects-by-AutoUpdate
```
Comment (stgraber) [inline lxd/db/images.go:32]: Should split this commit in half, one half for this comment + result of the code generator and then a second one for the manually added function.

File: shared/util.go [hunk 6]
```diff
@@ -1237,3 +1239,21 @@ func JoinUrls(baseUrl, p string) (string, error) {
 	u.Path = path.Join(u.Path, p)
 	return u.String(), nil
 }
+
+// GetStableRandomValueFromList returns a stable random value from a given list.
```
Comment (stgraber) [inline shared/util.go:1243]: @tomponline can we use that for the other cases you already implemented?

Comment (tomponline) [reply shared/util.go:1243]: @stgraber it looks like it could be used in `autoCreateCustomVolumeSnapshotsTask` and `addChassisGroupEntry`, as they pick from lists. However the other two use-cases are for generating MAC addresses. Searching for `fnv.New64a` shows all the current use-cases.

Also the seed in the other use-cases is a string not a single int.

The common parts are:

```go
hash := fnv.New64a()
_, err := io.WriteString(hash, seedStr)
if err != nil {
    return err
}

r := mrand.New(mrand.NewSource(int64(hash.Sum64())))
```

If this was refactored into two functions, firstly `GetStableRandomGenerator(seed string) *rand.Rand ` that just took a string seed and returned a preconfigured random generator, then it could be used for all use-cases, and then optionally a 2nd function `GetRandomNumber(r *rand.Rand, max int64) int64` this could then be used as so:

```go
r, err := GetStableRandomGenerator("someSeed")
if err != nil {
     return err
}

fmt.Printf("%d", myList[GetRandomNumber(r, len(myList))])
```

File: lxd/db/images.mapper.go [hunk 7]
```diff
@@ -1,4 +1,3 @@
-//go:build linux && cgo && !agent
```
Comment (stgraber) [inline lxd/db/images.mapper.go:1]: @monstermunchkin you still need to undo this change.

**Comment by stgraber** [comment]
@tomponline can you take a look at this PR in general too?

File: shared/util.go (hunk 6 above)
Comment (tomponline) [reply shared/util.go:1243]: @monstermunchkin please could you move this part (below) into its own function called `GetStableRandomGenerator(seed string) *rand.Rand`

```
hash := fnv.New64a()
_, err := io.WriteString(hash, seedStr)
if err != nil {
    return err
}

return mrand.New(mrand.NewSource(int64(hash.Sum64())))
```

Then you can use it in `GetStableRandomValueFromList` and I can use it later in the other stable random number generator scenarios. Thanks

File: lxd/db/images.go [hunk 8]
```diff
@@ -1138,6 +1138,17 @@ WHERE images.fingerprint = ?
...
 
+// GetNodesWithImageAndAutoUpdate returns the addresses of online nodes which already have the image.
+func (c *Cluster) GetNodesWithImageAndAutoUpdate(fingerprint string, autoUpdate bool) ([]string, error) {
+	q := `
+SELECT DISTINCT nodes.address FROM nodes
+  LEFT JOIN images_nodes ON images_nodes.node_id = nodes.id
+  LEFT JOIN images ON images_nodes.image_id = images.id
+WHERE images.fingerprint = ? AND images.auto_update = ?
+	`
+	return c.getNodesByImageFingerprintAndAutoUpdate(q, fingerprint, autoUpdate)
```
Comment (tomponline) [inline lxd/db/images.go:1149]: @monstermunchkin @stgraber I can see this pattern of defining the query in the exported function and then passing it to the unexported function to actually run the query is already in use in this file, but is there a reason that the contents of `getNodesByImageFingerprintAndAutoUpdate` couldn't just be in this function as its not used elsewhere?

File: lxd/db/images.go [hunk 9]
```diff
@@ -1138,6 +1138,17 @@ WHERE images.fingerprint = ?
 	return c.getNodesByImageFingerprint(q, fingerprint)
 }
 
+// GetNodesWithImageAndAutoUpdate returns the addresses of online nodes which already have the image.
+func (c *Cluster) GetNodesWithImageAndAutoUpdate(fingerprint string, autoUpdate bool) ([]string, error) {
+	q := `
+SELECT DISTINCT nodes.address FROM nodes
+  LEFT JOIN images_nodes ON images_nodes.node_id = nodes.id
```
Comment (tomponline) [inline lxd/db/images.go:1145]: @monstermunchkin from what I can see we aren't interested in the nodes that dont have this image, so is there a reason for the use of LEFT JOIN rather than just a straight JOIN here?

File: lxd/db/node.go [hunk 10]
```diff
@@ -109,6 +109,23 @@ func (c *ClusterTx) GetNodeByName(name string) (NodeInfo, error) {
 	}
 }
 
+// GetNodeByID returns the node with the given ID.
+func (c *ClusterTx) GetNodeByID(nodeID string) (NodeInfo, error) {
```
Comment (tomponline) [inline lxd/db/node.go:113]: @monstermunchkin it feels like the `nodeID` argument should be an int64 here to align with the data type used by the database layer when returning the ID. Then if you have a string it can be converted outside of the function.

File: lxd/images.go [hunk 11]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
 }
 
 func autoUpdateImages(ctx context.Context, d *Daemon) error {
```
Comment (tomponline) [inline lxd/images.go:1117]: @monstermunchkin please could you expand the commit message from "Fix auto image updates" to explain what is being fixed, as I confess I'm not sure how to review it without knowing what was broken with it :) thanks

File: lxd/db/node.go (hunk 10 above)
Comment (monstermunchkin) [reply lxd/db/node.go:113]: @tomponline Actually, this function can be removed. I believe I previously needed this function but now it's not used anymore.

File: lxd/db/images.go (hunk 9 above)
Comment (monstermunchkin) [reply lxd/db/images.go:1145]: @tomponline I just copied it from the function above :)

File: shared/util.go (hunk 6 above)
Comment (monstermunchkin) [reply shared/util.go:1255]: Done.

File: lxd/db/images.go (hunk 8 above)
Comment (monstermunchkin) [reply lxd/db/images.go]: @tomponline I removed `getNodesByImageFingerprintAndAutoUpdate` and use `getNodesByImageFingerprint` instead which now takes a `*bool` as well.

**Comment by stgraber** [comment]
@tomponline can you do another pass and if happy, merge it?
The issue being fixed is https://github.com/lxc/lxd/issues/8436

I've done a quick pass and the test seems to match what we want, the internal API endpoints look fine but I'd appreciate another set of eyes on the images.go changes themselves.

File: lxd/db/images.go (hunk 9 above)
Comment (tomponline) [reply lxd/db/images.go:1145]: @monstermunchkin does it work as expected using just `JOIN` as I've tended to avoid using `LEFT JOIN` unless I specifically wanted the behaviour it brings (i.e return all left hand side rows and return null for the columns on the right hand side if they dont match the join condition), this way it makes it clear that we're not interested in the rows if they don't match, and for larger result sets there is likely to be a performance gain by not requiring sqlite to generate all the rows in the result set. I realise in this case the query's WHERE statement is filtering out the null rows by filtering on the `images` table, but it feels a bit counter-intuitive to specify a LEFT JOIN (indicating we want all rows) and then filtering them out :).

If it were an existing query I'd probably leave it alone unless specifically working on it, but as this is a new query if it works as desired with a normal JOIN then its probably worth using it.

File: lxd/images.go (hunk 11 above)
Comment (tomponline) [reply lxd/images.go:1117]: @monstermunchkin thanks for this, it really helps when reviewing as that gives a good concise summary of what I should look out for :)

File: lxd/db/images.go (hunk 9 above)
Comment (monstermunchkin) [reply lxd/db/images.go:1145]: It works as expected with just `JOIN`. I'll update the commit.

File: lxd/images.go [hunk 12]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+				if err != nil {
+					return err
+				}
+
+				nodeIDs = append(nodeIDs, nodeInfo.ID)
+
+				return nil
+			})
+			if err != nil {
+				return errors.Wrap(err, "Unable to retrieve cluster node information")
```
Comment (tomponline) [inline lxd/images.go:1161]: @monstermunchkin please could you use "cluster member" rather than "node" or "cluster node" in your log and error messages. @stgraber has mentioned to me on my PRs that we avoid using "node" as it can mean many things, where as "cluster member" is more specific and also easier to search for.

File: lxd/images.go [hunk 13]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+		return errors.Wrap(err, "Unable to retrieve image fingerprints")
 	}
 
-	for _, project := range projectNames {
-		err := autoUpdateImagesInProject(ctx, d, project)
+	for fingerprint, images := range imageMap {
+		nodes, err := d.cluster.GetNodesWithImageAndAutoUpdate(fingerprint, true)
 		if err != nil {
-			return errors.Wrapf(err, "Unable to update images for project %s", project)
+			logger.Debug("Error getting nodes", log.Ctx{"fingerprint": fingerprint})
```
Comment (tomponline) [inline lxd/images.go:1141]: @monstermunchkin same point as below about use of "nodes" rather than "cluster members" in log message.

Comment (tomponline) [reply lxd/images.go:1141]: Also here, if you're changing it, might be worth expanding the log message to say something like "Error getting cluster members for image auto update" to give the log message a bit more context (as otherwise apart from inferring it from fingerprint also logged, we wouldn't know the log message is related to images, or the specific operation on them).

File: lxd/images.go [hunk 14]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+			continue
```
Comment (tomponline) [inline lxd/images.go:1142]: If this scenario occurs and we skip handling the image because we cannot load it's nodes, will this be retried at some point (just for my knowledge).

File: lxd/images.go (hunk 12 above)
Comment (tomponline) [inline lxd/images.go:1161]: @monstermunchkin should this step be moved into a separate function that is combined with `GetNodesWithImageAndAutoUpdate` so if either of them fail, you can `continue` onto the next image like you do if `GetNodesWithImageAndAutoUpdate` fails for an image?

File: lxd/images.go [hunk 15]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+				return nil
+			})
+			if err != nil {
+				return errors.Wrap(err, "Unable to retrieve cluster node information")
+			}
+		}
+
+		// If multiple nodes have the image, select one to deal with it.
+		if len(nodeIDs) > 1 {
+			selectedNode, err := shared.GetStableRandomInt64FromList(len(images), nodeIDs)
```
Comment (tomponline) [inline lxd/images.go:1167]: @monstermunchkin actually now I come to this section, it would be better to only convert the node names to node IDs in this block (to avoid doing unnecessary queries in the scenario where there is a single node with the image). Although the point about continuing to the next image on failure still stands though.

File: lxd/images.go (hunk 14 above)
Comment (monstermunchkin) [reply lxd/images.go:1142]: It will be retried the next time the function is run.

File: lxd/images.go [hunk 16]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+			if err != nil {
+				logger.Debug("Failed to get stable random value", log.Ctx{"err": err})
```
Comment (tomponline) [inline lxd/images.go:1169]: @monstermunchkin same point about expanding this message to include a prefix/suffix indicating its part of the image auto update operation.

Also shouldn't this be a `log.Error()` after all it is an error happening that is affecting the normal operating scenario.

File: lxd/images.go (hunk 13 above)
Comment (tomponline) [reply lxd/images.go:1141]: This should be a logger.Error right?

File: lxd/images.go [hunk 17]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+				continue
+			}
+
+			if d.cluster.GetNodeID() != selectedNode {
```
Comment (tomponline) [inline lxd/images.go:1173]: @monstermunchkin maybe a little comment here explaining that the matching node will run the task and explain that this is to stop multiple nodes trying to update and distribute the image concurrently.

File: lxd/images.go [hunk 18]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+				continue
+			}
+		}
+
+		var newImage *api.Image
```
Comment (tomponline) [inline lxd/images.go:1178]: @monstermunchkin as `newImage` isn't used outside of the `go func() {}` function it would make sense to reduce its scope and define it inside the function.

Also, my understanding of the go routine here is so that it allows for cancellation of the ongoing update and distribution operation in case the LXD daemon is stopped and the task is cancelled.

However just returning nil immediately doesn't clean up the go routine (and we can't know that the reason the task is being cancelled is because the LXD daemon is shutting down, the context could be cancelled for other reasons unbeknownst to this function). Also, if LXD daemon is shutting down (perhaps due to snap update) and the go routine is finished unexpectedly then we risk killing an unfinished image update or distribution operation unexpectedly. I believe this could cause some inconsistency that we would be aiming to fix with this PR (but I may be going beyond the scope of the PR that @stgraber intended so if so no worries).

But assuming cancelling the update and distribution process cleanly is in scope, then rather than immediately returning from this function pass the context to `autoUpdateImage` and `distributeImage` so they can finish up cleanly, and then return the error from `ctx.Err()` from those functions when context is cancelled, which you can then check here for being `context.Canceled` and if so return from the function.

I'm not sure how difficult it is to update `autoUpdateImage` and `distributeImage` to cleanly finish up on context cancellation, but it feels like adding the plumbing for it at least to those functions is worth doing now, even if they just have a comment saying "//TODO handling context cancellation properly" for now.

File: lxd/images.go [hunk 19]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+			logger.Errorf("Failed to retrieve the information of node: %v", err)
+			return err
 		}
-	} else {
-		interval, err = cluster.ConfigGetInt64(d.cluster, "images.auto_update_interval")
+
+		client, err := cluster.Connect(nodeAddress, d.endpoints.NetworkCert(), true)
 		if err != nil {
-			return errors.Wrap(err, "Unable to fetch cluster configuration")
+			return errors.Wrap(err, "Failed to connect node for image synchronization")
```
Comment (tomponline) [inline lxd/images.go:1265]: @monstermunchkin would be useful to say which node in the error context using `%q` and `nodeAddress` (assuming its not in the `err` already that is).

File: lxd/images.go [hunk 20]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+
+		var nodeInfo db.NodeInfo
+		err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+			var err error
+			nodeInfo, err = tx.GetNodeByAddress(nodeAddress)
+			return err
+		})
 		if err != nil {
-			return errors.Wrap(err, "Unable to fetch project configuration")
+			logger.Errorf("Failed to retrieve the information of node: %v", err)
```
Comment (tomponline) [inline lxd/images.go:1259]: @monstermunchkin adding the node address here would be good, could use logger.Errorf("Failed to retrieve the information of cluster member %q: %v", nodeAddress)

File: lxd/images.go [hunk 21]
```diff
@@ -1115,93 +1115,214 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+		op, err := client.CreateImage(image, createArgs)
 		if err != nil {
-			logger.Error("Error loading image", log.Ctx{"err": err, "fingerprint": fingerprint, "project": projectName})
-			continue
+			return err
 		}
 
-		if !info.AutoUpdate {
-			continue
+		err = op.Wait()
```
Comment (tomponline) [inline lxd/images.go:1302]: @monstermunchkin maybe the op can be cancelled if the context is cancelled?

File: lxd/images.go [hunk 22]
```diff
@@ -2352,7 +2499,8 @@ func imageRefresh(d *Daemon, r *http.Request) response.Response {
 
 	// Begin background operation
 	run := func(op *operations.Operation) error {
-		return autoUpdateImage(d, op, imageId, imageInfo, project)
```
Comment (tomponline) [inline lxd/images.go:2355]: @monstermunchkin is `imageRefresh` even needed anymore?

**Comment by tomponline** [comment]
@monstermunchkin something I noticed in reviewing this is that the comment here https://github.com/lxc/lxd/blob/master/lxd/daemon.go#L1232 doesn't match with the `task.Every(time.Hour)` in `autoUpdateImagesTask`, so not sure which is correct, but worth reviewing and updating while you're looking at image auto updates.

File: lxd/images.go (hunk 12 above)
Comment (monstermunchkin) [reply lxd/images.go:1161]: Personally, I don't see the necessity of moving this to a separate function. Also, I can't think of a good name for that function :D

Comment (tomponline) [reply lxd/images.go:1161]: @monstermunchkin yeah I changed my mind on this later once I found out it was only used inside the block if there is >1 nodes https://github.com/lxc/lxd/pull/8403#discussion_r587334580

File: lxd/images.go (hunk 22 above)
Comment (monstermunchkin) [reply lxd/images.go:2355]: @tomponline No. This is for testing only.

Comment (tomponline) [reply lxd/images.go:2355]: So it can be removed? I couldnt see references to it in the code base.

Comment (monstermunchkin) [reply lxd/images.go:2355]: It's used in our test suite to trigger an image refresh.

Comment (tomponline) [reply lxd/images.go:2355]: @monstermunchkin ah ok, yes I just found it, was searching for `imageRefresh(` but its used as a handler in `imageRefreshCmd`

**Comment by stgraber** [comment]
@tomponline the default is indeed every 6 hours, so that comment is correct. We run the task every hour to handle projects being able to override the frequency, so every hour we look at which one need to be updated.

**Comment by tomponline** [comment]
> @tomponline the default is indeed every 6 hours, so that comment is correct. We run the task every hour to handle projects being able to override the frequency, so every hour we look at which one need to be updated.

Cool thanks, would be good to add a comment there to explain it to future readers.

File: lxd/images.go [hunk 23]
```diff
@@ -1115,93 +1115,234 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+		return errors.Wrap(err, "Unable to retrieve image fingerprints")
 	}
 
-	for _, project := range projectNames {
-		err := autoUpdateImagesInProject(ctx, d, project)
+	for fingerprint, images := range imageMap {
+		nodes, err := d.cluster.GetNodesWithImageAndAutoUpdate(fingerprint, true)
 		if err != nil {
-			return errors.Wrapf(err, "Unable to update images for project %s", project)
+			logger.Error("Error getting cluster members", log.Ctx{"fingerprint": fingerprint})
```
Comment (stgraber) [inline lxd/images.go:1141]: Per @tomponline initial comment, you should change that to `Error getting cluster members for image auto-update` so if it shows up in logs we can figure it out.

File: lxd/images.go [hunk 24]
```diff
@@ -1115,93 +1115,234 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+					deleteIDs = append(deleteIDs, image.ID)
+				}
+
+				if newImage == nil {
+					newImage = newInfo
+				}
+			}
+
+			if newImage != nil {
+				err := distributeImage(ctx, d, nodes, fingerprint, newImage)
```
Comment (tomponline) [inline lxd/images.go:1211]: @monstermunchkin I was more thinking of passing ctx into `autoUpdateImage` and `distributeImage`, and updating the handling here to detect for them returning the `context.Canceled` error and returning at that point rather than using the go routine here. And then inside those two functions adding a comment about needing to handle context properly. That way the plumbing for it is there and this function won't return until those ones have returned cleanly.

That way this function won't have to use ctx.Done() at all.

File: lxd/images.go (hunk 18 above)
Comment (stgraber) [reply lxd/images.go]: @monstermunchkin I made the same comment on Tuesday, can you re-introduce the comment that this PR removed?

It used to say:
```
		// FIXME: since our APIs around image downloading don't support
		//        cancelling, we run the function in a different
		//        goroutine and simply abort when the context expires.
```

File: lxd/images.go (hunk 24 above)
Comment (monstermunchkin) [reply lxd/images.go]: @tomponline Like this? (see last commit)

**Comment by stgraber** [comment]
@monstermunchkin still got that one missing comment mentioned above to sort out.

**Comment by monstermunchkin** [comment]
> @monstermunchkin still got that one missing comment mentioned above to sort out.

@stgraber The latest commit removes the go routine. Therefore the comment isn't needed.

**Comment by stgraber** [comment]
Ah, good :)

File: lxd/images.go [hunk 25]
```diff
@@ -1115,93 +1115,223 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+						return err
+					}
+
+					nodeIDs = append(nodeIDs, nodeInfo.ID)
+
+					return nil
+				})
+				if err != nil {
+					logger.Error("Unable to retrieve cluster member information for image update", log.Ctx{"err": err})
+					continue
```
Comment (tomponline) [inline lxd/images.go:1167]: @monstermunchkin I think we need to skip the whole image update if this happens, not just the retrieving the node info, because otherwise we may end up incorrectly thinking there are only 1 nodes and then handling the update on more than 1 nodes at once.

**Comment by monstermunchkin** [comment]
@tomponline you mean like this?
```go
func distributeImage(ctx context.Context, d *Daemon, nodes []string, oldFingerprint string, newImage *api.Image) error {
	ch := make(chan (struct{}))

	go func() {
		// here goes the code
		// ...
		// done
		ch <- struct{}{}
	}()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-ch:
		return nil
	}
}

func autoUpdateImages(ctx context.Context, d *Daemon) error {
	// ...
	err := distributeImage(ctx, d, nodes, oldFingerprint, newImage)
	if err != nil {
		logger.Error("foo")
		
		if err == context.Canceled {
			return nil
		}
	}

	return nil
}
```

If I understand go correctly, this doesn't make sense since the goroutine will continue to run even if the context is cancelled.

**Comment by tomponline** [comment]
@monstermunchkin yeah that looks right.

**Comment by monstermunchkin** [comment]
> @monstermunchkin yeah that looks right.

The goroutine will still continue to run even if the context is cancelled. That seems wrong to me.

**Comment by tomponline** [comment]
> > @monstermunchkin yeah that looks right.
> 
> The goroutine will still continue to run even if the context is cancelled. That seems wrong to me.

Yeah that was the original problem, but unless the functions you're calling inside `distributeImage` also support context cancellation, then there's not much we can do here.

**Comment by monstermunchkin** [comment]
> > > @monstermunchkin yeah that looks right.
> > 
> > 
> > The goroutine will still continue to run even if the context is cancelled. That seems wrong to me.
> 
> Yeah that was the original problem, but unless the functions you're calling inside `distributeImage` also support context cancellation, then there's not much we can do here.

I agree. However, we still don't need goroutines for that. The last commit checks `ctx.Done()` in one of the loops and returns if the context has been cancelled.

File: lxd/images.go [hunk 26]
```diff
@@ -1115,93 +1115,226 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+			} else {
+				deleteIDs = append(deleteIDs, image.ID)
+			}
+
+			if newImage == nil {
+				newImage = newInfo
+			}
+		}
+
+		if newImage != nil {
```
Comment (tomponline) [inline lxd/images.go:1212]: @monstermunchkin should this block be inside the `for _, image := range images {` block above, so its done for each image?

File: lxd/images.go [hunk 27]
```diff
@@ -1115,93 +1115,226 @@ func autoUpdateImagesTask(d *Daemon) (task.Func, task.Schedule) {
...
+			err := distributeImage(ctx, d, nodes, fingerprint, newImage)
+			if err != nil {
+				logger.Error("Failed to distribute image", log.Ctx{"err": err, "fingerprint": newImage.Fingerprint})
+
+				if err == context.Canceled {
+					return nil
+				}
+			}
+
+			for _, ID := range deleteIDs {
```
Comment (tomponline) [inline lxd/images.go:1222]: @monstermunchkin for non-nil errors, should we still try and delete images here?

File: lxd/images.go (hunk 26 above)
Comment (monstermunchkin) [reply lxd/images.go:1212]: @tomponline Nope. `distributeImage` should only be called once for each fingerprint. `images` is a list of images with the same fingerprint but e.g. different node ID or project ID.

Comment (tomponline) [reply lxd/images.go:1212]: @monstermunchkin does it not matter if `autoUpdateImage` could return several `newInfo` over its different executions?

It looks a little counter-intuitive there, so a comment explaining that would be nice.

File: lxd/images.go (hunk 27 above)
Comment (monstermunchkin) [reply lxd/images.go:1222]: @tomponline Oh yeah, I missed that.

File: lxd/images.go (hunk 26 above)
Comment (monstermunchkin) [reply lxd/images.go:1212]: @tomponline `newInfo` will basically always be the same for each image in the loop. Therefore, we just pick the first one. I'll add a comment.

File: lxd/images.go (hunk 27 above)
Comment (monstermunchkin) [reply lxd/images.go:1224]: @tomponline I had to revert that. It is important that `newInfo` is non-nil. Otherwise the old image will be deleted even if the image has not been updated.

Comment (tomponline) [reply lxd/images.go:1224]: @monstermunchkin if distributeImage returns a non-nil error, should we still try to DeleteImage (as it is now) or should we skip it?

Comment (monstermunchkin) [reply lxd/images.go:1224]: @tomponline We should still delete the old images.

File: lxd/api_internal.go [hunk 28]
```diff
@@ -108,6 +112,49 @@ var internalRAFTSnapshotCmd = APIEndpoint{
...
+	err = imageCreateInPool(d, &req.Image, req.Pool)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	return response.EmptySyncResponse
+}
+
+func internalRefreshImage(d *Daemon, r *http.Request) response.Response {
+	err := autoUpdateImages(context.Background(), d)
```
Comment (tomponline) [inline lxd/api_internal.go:150]: @monstermunchkin shouldnt this be the passed the context from the daemon rather than the base context, in case the daemon itself is shutdown?

**Review by tomponline — APPROVED** [review]

**Comment by stgraber** [comment]
@monstermunchkin that's not looking good...
